### ISSUE TITLE
Proper mutability in N

### ISSUE THREAD

**Issue by SheepTester** [nbuilding/N-lang#166]
## The problem

Currently, all variables can be changed using `var`, which is very unsafe. Oftentimes we write code assuming our variables won't suddenly change; for example, we assume that `intInBase10` will always be the same `intInBase10`. However, if any variable can be `var`'d, this assumption can no longer be made.

Also, `var`'s behaviour is ill-defined in some edge cases like classes, records, and function arguments, and requires that the execution order of presumably pure functions be defined. For example, it should not matter whether a third party tool peeks into a function's return value or not.

Here's why mutable variables (and impure functions) suck:

- When debuggng, printing an impure function to see its output can unexpectedly change the behaviour the program.
- Third party tools cannot preview or peek into return values of functions without causing some other side effect.
- Async tasks running in parallel become unstable. For example, `var count = count + 1` might be an issue if, while adding `count + 1`, another thread sets `count` to a new value. Then, `count + 1` reflects a sum with the old value and is set to `count`, which overrides the first change.

## Possible solution: `let var`

For backwards compatibility, existing `var` statements will just emit warnings. This is fine; the JS branch already intended on marking `var` as undefined behaviour.

Perhaps we can do

```ts
let var mutableVariable = { count: 0 }
```

where `var` here indicates that the variable is mutable (or variable, as opposed to constant).

Then, the variable is free to be mutated without warnigns:

```ts
var mutableVariable = { count: mutableVariable.count + 1 }
```

`mutableVariable` is usable like a normal immutable value:

```ts
let immutable: { count: int } = mutableVariable
```

This makes a *deep copy* of the mutable value, so changing `mutableVariable` won't affect the value of `immutable`.

Now, let's take an example which currently has undefined/ill-defined behaviour:

```ts
let makeCounter = [_: ()] -> () -> int {
  let count = 0
  return [_: ()] -> int {
    var count = count + 1
    return count
  }
}

let counter = makeCounter(())
print(counter(())) // 1
print(counter(())) // 2
print(counter(())) // 3
```

`counter` violates expectations. Despite giving the same value `()`, it returns a different value each time. This will harm debugging; trying to print a function call might change the behaviour of a program.

One solution would be to use a `cmd`. However, the problem with `cmd`s is that they can only be used inside other `cmd` functions. A function using `makeCounter` can still be pure. For example,

```ts
let addFirstIntegers = [upTo: int] -> int {
  let sum = 0
  let counter = makeCounter(())
  for (_ in range(0, upTo, 1)) {
    var sum = sum + counter(())
  }
  return sum
}
```

Despite using `var` and an impure function, on the outside, `addFirstIntegers` is pure. Giving it the same value for `upTo` will always make it return the same output.

Ideally, there should be a way to denote functions that are impure but can be safely used inside pure functions. Perhaps this can be done as a fancy `var` mutable type, like `var int` for a mutable int.

Another case might be to pass mutable pointers to other functions. For example, we might want a `remove` function that modifies a mutable list.

```ts
let var mutableList = [1, 2, 3]
removeItem(1, mutableList) // Removes second item (index 2) from list
```

With what I've said above, this currently would make a copy of `mutableList`, which is slow and won't mutate the `mutableList`. Ideally, `removeItem` should be able to modify `mutateList` from inside the function.

## Summary of issues

`let var` solves one problem, but I list two more problems that it can't solve alone:

1. Denoting impure functions that can be used in a pure function (such as `counter`)
2. Making functions that can mutate a given mutable value

Let's discuss some possible solutions to those or alternatives to `let var` here. Hopefully with these improvements to mutability, we can make N a much more pleasant and reliable language to use.

**Comment by Ashvin-Ranjan** [comment]
The main problem with this is that we would need to make a way of keeping this backward compatible, maybe instead we can make `const` and switching the standard to that?

**Comment by SheepTester** [comment]
> The main problem with this is that we would need to make a way of keeping this backward compatible, maybe instead we can make `const` and switching the standard to that?

We can make `var` raise a warning rather than an error for immutable variables

**Comment by SheepTester** [comment]
We could eliminate the first issue by rewriting `makeCounter`. Maybe if a function takes a `var int`, it takes a *reference* to a mutable variable of type `int`. For example,

```ts
let makeCounter = [_: ()] -> (var int -> int, int) {
  let state = 0
  return [state: var int] -> int {
    var state = state + 1
    return state
  }, state
}

let counter, var state = makeCounter(())
print(counter(state)) // prints 1, state = 1
print(counter(state)) // prints 2, state = 2
print(counter(state)) // prints 3, state = 3
```

For this example it's obviously overcomplicated, but you can imagine how a library can put all its state in an aliased record type so that the outer user

**Comment by SheepTester** [comment]
There are a lot of functions for lists like `append` `prepend` `remove` that could use mutable arguments, but it would be rather inconvenient to chain them if they don't return any useful value. Ideally, both of the following should do the same thing:

```ts
let var temp: list[str] = []
append("wow", temp)
append("weee", temp)
prepend("wow2", temp)
remove(1, temp)
let list = temp // ["wow2", "weee"]
```

```ts
let list = []
  |> append("wow")
  |> append("weee")
  |> prepend("wow2")
  |> remove(1)
```

The latter case demonstrates that `append` etc. should return a list, and the former case demonstrates that the functions should take a `var list[str]`. Perhaps, then, `append`'s type should be `[t] t -> var list[t] -> list[t]`

This suggests that for the latter case, passing in a `list[t]` to a function that takes `var list[t]` will make a copy (or at least, it won't mutate the original list). This is inefficient, though; why should the list be cloned at every step?

Ideally, because the list remains the same from one step to the next, internally, the N executor should use the same list in memory for each step. Maybe this can just be an optimisation, though. After all, if `append` returns the mutable list, it won't be mutated any more since the function has ended.

**Comment by SheepTester** [comment]
I also wonder if it is possible to do something like `(int, var str) -> bool` or `str -> var bool`. What would they mean? How should the grammar or type checker prohibit weird uses of `var`? If `var` is like a type operator, then it'd be possible to do (at least in the syntax)

```ts
let wow: var var var var var var var var var str = "weee"
```

I think ``(int, var str) -> bool`` makes sense, but it implies that there's a mutable reference inside an immutable tuple, which seems not so safe. Not sure. I think the entire tuple should also be mutable, but then 

`str -> var bool` could make sense if the function declares a mutable variable of type `var` then returns a mutable reference to it. However, does it make a difference whether it's `var bool` or `bool`? Since the function has ended, it can't mutate its `bool` variable anymore, so it might as well be immutable. Thus, it'd just be an optimisation to return the mutable value straight up regardless of whether it returns `bool` or `var bool` (like Rust, transferring ownership out of a function).

Also, maybe it's possible that you want to use a function that takes a mutable reference on a *clone* of a mutable value. I guess one way would be to use an identity function since it doesn't have `var` so it has to clone. This also means that `print`'s type signature would have to change to `[t] var t -> t` if it were a true identity function. Hmm, idk.

Though I guess come to think of it, if a function returns a mutable reference from its arguments, then that mutable reference can still be changed afterwards. 😳

**Comment by SheepTester** [comment]
Hmm, but it'd still be nice to tell whether `someFunction(value)` is pure or not without having to look up whether `value` and `someFunction` both are var

**Comment by SheepTester** [comment]
> Hmm, but it'd still be nice to tell whether `someFunction(value)` is pure or not without having to look up whether `value` and `someFunction` both are var

Maybe we could use pointers instead or something, idk, like `someFunction(&value)`

**Comment by Ashvin-Ranjan** [comment]
Pointers are scary, very very scary

**Comment by SheepTester** [comment]
Well, Python has them too; lists and other data types are usually passed around by reference rather than cloning everything

**Comment by Ashvin-Ranjan** [comment]
True

**Comment by Ashvin-Ranjan** [comment]
I think that this is the way to go. For functions, we can do `int -> var str -> bool`, so you can mutate the value of the string in the function.

**Comment by SheepTester** [comment]
Actually, maybe instead of `let var`, we can require a type annotation on mutable variables and then move `var` to the type annotation, like

```ts
let wow: var str = "sheep"
var wow = "wiggle"
```

Requiring a type annotation is so we know what goes inside an empty list (`let list: var list[str] = []`)

**Comment by Ashvin-Ranjan** [comment]
So instead of variable modifiers `var` will be a type modifier?

**Comment by SheepTester** [comment]
Kind of, hmm.

Maybe I'll also make it required to use the `var` keyword whenever passing around the mutable reference.

So mutable values represent the same values in memory, so

```ts
let a: var list[int] = [1]
let b: var list[int] = var a // Simply `a` would clone the list
append(2, var b) // Simply `b` would clone the list, so the function would return a new list without modifying b
print(a) // Clones `a`, which is fine; giving `var a` would be a type error. Prints [1, 2]
```

Actually, requiring type annotations is still quite cumbersome. Maybe we should still have `let var` as a shorthand.

Also, a value of type `t` is assignable to `var t`, making a clone of the value, but `var t` is not assignable to `t`.

Would something like `var [1]` make sense? Or should `var` as an operator be limited to variable names?

Also, it might be surprising behaviour if a user forgets to add `var` in `append(2, var b)`. Maybe cloning a `var` value could be explicit, like only either `append(2, var b)` or `append(2, clone b)` and disallow `append(2, b)` if the function takes a var value. Explicitly saying `var` helps indicate that the function has side effects.

To implement a function `appendTwice`, which appends a value twice, it could be defined as

```ts
let appendTwice = [[t] item:t list:var list[t]] -> var list[t] {
  append(item, var list)
  append(item, var list)
  return var list // Returns the mutable reference to the given list, for chaining with pipes (to avoid unnecessary cloning)
}
```

So even if a variable like `a` or `list` in the above examples are mutable, you still explicitly have to do `var a`/`var list`; `a` on its own is not acceptable. Hmm

What about function type variables? `print : [t] t -> t` implies that `print` should be able to take a mutable value, and it returns another mutable value, though not necessarily the same mutable reference in memory. I suppose it works

What about currying/functions returning functions? `str -> int` should always return the same `int` given the same `str`, but if it came from giving a function `var list[str] -> str -> int`, then that might not be the case:

```ts
let badAppend = [list:var list[str] item:str] -> int {
  append(item, var list)
  return len(list)
}

let var myList = ["hello"]
print(badAppend(var myList, "hi")) // 2

let adder: str -> int = badAppend(var myList)
print(adder("wow")) // 3
print(adder("wow")) // 4??
```

This is terrible! I think the solution here could be that functions cannot take a mutable type if they return a function

**Comment by SheepTester** [comment]
Okay, so I guess here's a summary of the proposal so far:

- We're introducing a new type of types: `var` mutable types
  - In type annotations, it's an operator you can apply to any type
    - **TODO**: Is it fine for a mutable type to be inside an immutable type, like `(str, var list[int])`?
    - In `let` statements, you can take advantage of type inference by using `let var` to infer a type and then make it mutable. For example, `let var wow = [1, 2, 3]` is the same as `let wow: var list[int] = [1, 2, 3]`
      - **NOTE**: Presumably this means you cannot specify a type annotation if you use `let var`?
      - **TODO**: Is it an error if a type such as `[]`, `none`, or `1` cannot be immediately and unambiguously inferred alone, or can we figure it out from later usage?
  - Mutable values can be passed around like their immutable counterpoints.
    - To pass a value of type `var t` to what should be a `var t` (e.g. as a function argument or in a `let` statement), you must use the `var` keyword to indicate you're passing along a mutable reference and it could be changed.
      - **TODO**: This requirement might be unnecessarily verbose. I might rethink this.
    - To pass a value of type `var t` to what should be a `t` (passing a mutable value as an immutable value), it must be cloned explicitly. Probably this can just be a global function `clone`.
  - Functions that take a mutable type cannot return a function. This also means that the mutable type can only be used as the last argument of a function expression.
  - We continue to uphold the idea that a function of immutable types should remain pure.

**Comment by Ashvin-Ranjan** [comment]
I think we should allow casting between mutable and immutable types.

**Comment by ky28059** [comment]
My opinions:
- I think it's easier to have two keywords for mutability (like kotlin's `var`/`val` or js's `let`/`const`) instead of `let var` which is cumbersome
- Would `var list[int]` mean you can modify the list itself? In kotlin `var x: List<t>` just means you can reassign it to another list; if you wanted to do `x[0] = blah` you'd need a `MutableList<t>`
- On passing a value of `t` where `var t` is needed, how would that work / when would that be used? I'm assuming something like 
```js
function foo(var t: int) { }
```
seeks to mutate t, which is why it explicitly requires var (otherwise `t: int` would work fine). If this is the case, you cannot pass `t` into something that requires `var t`, as that would violate mutability.
```js
let var a = 3
let b = 3
function foo(var t: int) { t += 3 }
foo(a) // fine
foo(b) // not good
```
Otherwise, if passing a `t` into a `var t` doesn't mutate `t`, what's the point of asking for a `var t` then?

**Comment by Ashvin-Ranjan** [comment]
> * I think it's easier to have two keywords for mutability (like kotlin's `var`/`val` or js's `let`/`const`) instead of `let var` which is cumbersome

I do feel like it is better to make a `const` keyword instead of adding a modifier now that you say it

> * Would `var list[int]` mean you can modify the list itself? In kotlin `var x: List<t>` just means you can reassign it to another list; if you wanted to do `x[0] = blah` you'd need a `MutableList<t>`

Lists are going to be immutable, the way that we currently assume someone would set an item in a list is using the upcoming `..` operator, as you cannot set an item in a list.

For your last point I am still unsure of how we should go about it

**Comment by ky28059** [comment]
> For your last point I am still unsure of how we should go about it

The way kotlin does it is simply not allow you to mutate values (at least to my knowledge). All function parameters in kotlin are immutable `val`s and they are passed by value instead of reference.
```kt
fun foo(x: Int) {
    x += 3 // illegal: val cannot be reassigned
}
```
The way c# does it is that all variables are mutable, but are still passed by reference unless you use one of the [proper keywords](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier).
```cs
int x = 5;
void foo(out int x) {
    x = 3;
}
foo(out x);
Console.WriteLine(x); // 3
```
The behavior of the c# out parameter keyword is what I assumed `var t` was trying to accomplish. I think adding `var t` is fine, as long as you don't allow passing in immutable `t`s to something that requires `var t`.

**Comment by Ashvin-Ranjan** [comment]
Yeah, maybe we can use `var` instead of out in this situation.

**Comment by SheepTester** [comment]
> I do feel like it is better to make a `const` keyword instead of adding a modifier now that you say it

Most languages except for JavaScript use `let` for constants, and `let` is already used for constants in existing versions of N. I think we could find a new keyword for mutable variables. I believe Rust does `let mut` and Swift uses `var`



> * Would `var list[int]` mean you can modify the list itself? In kotlin `var x: List<t>` just means you can reassign it to another list; if you wanted to do `x[0] = blah` you'd need a `MutableList<t>`

Yes. Currently in N if you do `let x: list[t] = ...`, `x` can't be reassigned nor its contents modified. I don't think there's an observable difference (other than performance) between reassigning and changing the contents of a list if the list binding itself is mutable.

Since these are mutable references, you can do something like

```ts
let changeToWow = [string: var str] -> () {
  var string = "wow"
}

let var happy = "happy"
changeToWow(var happy)
print(happy) // "wow"
```

In the future, mutable strings can also be used like a StringBuilder in Java for performance reasons

> * On passing a value of `t` where `var t` is needed, how would that work / when would that be used?

Some functions like `append` might both take and return a mutable reference to a list (for chaining function calls). That way, you can pass in an immutable list and use its return type:

```ts
let someFunction = [numbers: list[int]] -> ... {
  let newList = append(0, numbers)
  ...
}
```

> If this is the case, you cannot pass `t` into something that requires `var t`, as that would violate mutability.

If the function requires `var t`, N can implicitly clone `t` so that the function can mutate the cloned `t` while preserving the original `t`

Hmm, I suppose this might cause some silent errors if someone accidentally passes an immutable value to a function expecting it to change. Currently `var` is required to pass a mutable reference to a function, so that would raise an error if done on an immutable variable, but if the programmer also forgets to do that, then it becomes a silent error

> Lists are going to be immutable, the way that we currently assume someone would set an item in a list is using the upcoming `..` operator, as you cannot set an item in a list.

Immutable by default; with this proposal, we'll be able to allow lists to be mutated such as setting an item if the list is mutable

**Comment by ky28059** [comment]
> If the function requires `var t`, N can implicitly clone `t` so that the function can mutate the cloned `t` while preserving the original `t`

Hmm, I still think it's still better to explicitly clone the immutable variable before using it as a mutable variable.
```kt
val x = 3
fun whatever(x: Int): Int {
    var sum = x
    while (sum < 100) sum *= 2
    return sum
}
println(whatever(x))
```
> Some functions like `append` might both take and return a mutable reference to a list (for chaining function calls). That way, you can pass in an immutable list and use its return type

Not sure what this means. Could append not take and return a value instead?
```kt
fun append(elem: Int, arr: List<Int>): List<Int> {
    return listOf(*arr, elem)
}
```

**Comment by SheepTester** [comment]
`append` doesn't need a mutable reference to the list indeed, but for performance reasons it's probably preferable to just mutate the original list instead of `var list = append(item, list)`, which would create an entire new list

Explicit cloning is fine sure

**Comment by Ashvin-Ranjan** [comment]
@SheepTester Should this be closed with the creation of #218
